### ISSUE TITLE
Exception in callback _SelectorDatagramTransport._read_ready()

### ISSUE THREAD

**Issue by BettySwallocks** [rospogrigio/localtuya#1588]
My logs are getting filled with errors (every few seconds) like this:

```
2023-12-15 15:07:45.646 ERROR (MainThread) [homeassistant] Error doing job: Exception in callback _SelectorDatagramTransport._read_ready()
Traceback (most recent call last):
File "/config/custom_components/localtuya/discovery.py", line 65, in datagram_received
data = decrypt_udp(data)
^^^^^^^^^^^^^^^^^
File "/config/custom_components/localtuya/discovery.py", line 30, in decrypt_udp
return _unpad(decryptor.update(message) + decryptor.finalize()).decode()
^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 186, in finalize
data = self._ctx.finalize()
^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/cryptography/hazmat/backends/openssl/ciphers.py", line 222, in finalize
raise ValueError(
ValueError: The length of the provided data is not a multiple of the block length.
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
self._context.run(self._callback, *self._args)
File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 1163, in _read_ready
self._protocol.datagram_received(data, addr)
File "/config/custom_components/localtuya/discovery.py", line 67, in datagram_received
data = data.decode()
^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xd2 in position 4: invalid continuation byte
```

Any ideas folks?

**Comment by kozfelipe** [comment]
What are the names of discovered devices? My guess is one of them has invalid characters

**Comment by BettySwallocks** [comment]
> What are the names of your discovered devices? My guess is one of them has invalid characters at the name / attributes

Standard alphanumeric names - nothing notable, no special characters at all:

![image](https://github.com/rospogrigio/localtuya/assets/13366999/763bc8c3-1af1-46aa-9597-6551026c59b9)

**Comment by kozfelipe** [comment]
look up for invalid characters at the configuration attributes too

**Comment by BettySwallocks** [comment]
There are no odd characters in any attributes or other areas of config that I can find.  Mostly there is no complexity in my basic Tuya plug items, just two CT clamp style meters that have more attributes:

```
type: entities
entities:
  - entity: sensor.ltdm_ac_freq
    name: AC Freq
    secondary_info: last-updated
  - entity: sensor.ltdm_current_a
    name: Hob Current
  - entity: sensor.ltdm_current_b
    name: Ovens Current
  - entity: sensor.ltdm_energy_forward_a
    name: Hob Energy Forward
  - entity: sensor.ltdm_energy_forward_b
    name: Ovens Energy Forward
  - entity: sensor.ltdm_energy_reverse_a
    name: Hob Energy Reverse
  - entity: sensor.ltdm_energy_reverse_b
    name: Ovens Energy Reverse
  - entity: sensor.ltdm_power_a
    name: Hob Power
  - entity: sensor.ltdm_power_b
    name: Ovens Power
  - entity: sensor.ltdm_voltage
    name: Voltage
title: Hob Oven Meter
```

```
type: entities
entities:
  - entity: sensor.hp_imm_meter_current_a
    name: Current A
  - entity: sensor.hp_imm_meter_current_b
    name: Current B
  - entity: sensor.hp_imm_meter_power_a
    name: Power A
  - entity: sensor.hp_imm_meter_power_b
    name: Power B
    secondary_info: last-updated
  - entity: sensor.hp_imm_meter_today_energy_a
    name: Today energy A
  - entity: sensor.hp_imm_meter_today_energy_b
    name: Today energy B
  - entity: sensor.hp_imm_meter_total_energy
    name: Total energy
  - entity: sensor.hp_imm_meter_total_energy_a
    name: Total energy A
  - entity: sensor.hp_imm_meter_total_energy_b
    name: Total energy B
  - entity: sensor.hp_imm_meter_voltage_a
    name: Voltage A
  - entity: sensor.hp_imm_meter_voltage_b
    name: Voltage B
title: Heat Pump & Immersion Meter
```

If you're talking about the Tuya IOT Cloud conifgs - again I don't see anything in there with odd characters bar one of the meters which does have come Chinese characters in the "Product name" value, but that's been there for months.

**Comment by kozfelipe** [comment]
not the cloud configs, confirm there is no odd characters at the localtuya config

![image](https://github.com/rospogrigio/localtuya/assets/25208457/ab6c1648-9a5d-45fc-9719-43f7b478ee2f)

**Comment by BettySwallocks** [comment]
OK - I may not be looking in exactly the right way here, but here is my dual meter used for hob and oven monitoring:

![image](https://github.com/rospogrigio/localtuya/assets/13366999/f03fea2d-807a-4d0e-bfd4-009aa86e66d7)

![image](https://github.com/rospogrigio/localtuya/assets/13366999/a8353a46-3d5f-47e8-b523-f90253a940d9)

Is this the area to check?

**Comment by BettySwallocks** [comment]
Other than above my other items are single attribute plugs.
I have version 3.3 used for my plus and did use version 3.4 for the dual meter - anything wrong with that?

**Comment by kozfelipe** [comment]
my guess is that your discovery process is failing to read an invalid datagram from one of your devices according to the logs

**Comment by BettySwallocks** [comment]
Thanks for all the help.

That one meter is the only complex item in my config, so I'm guessing it's that as each of the other four are just plugs with simple on-off switch.  I may remove those plugs as they aren't actually used for anything as they are in normal Tuya integration also.

As for the meter I just didn't want to completely reset that as it's used in my energy statistics in HA. Hoping that it can be wiped and re-installed with the same name and then the stats will be picked back up...

**Comment by BettySwallocks** [comment]
Interesting, I've taken out all the plugs AND the meter so no devices are linked to the integration.  I've waited for 5 minutes and done a reload of the LocalTuya integration also.  The error in the logs is still regularly repeating:

```
This error originated from a custom integration.

Logger: homeassistant
Source: custom_components/localtuya/discovery.py:67
Integration: LocalTuya integration (documentation, issues)
First occurred: 15 December 2023 at 15:07:45 (74581 occurrences)
Last logged: 22:45:45

Error doing job: Exception in callback _SelectorDatagramTransport._read_ready()
Traceback (most recent call last):
  File "/config/custom_components/localtuya/discovery.py", line 65, in datagram_received
    data = decrypt_udp(data)
           ^^^^^^^^^^^^^^^^^
  File "/config/custom_components/localtuya/discovery.py", line 30, in decrypt_udp
    return _unpad(decryptor.update(message) + decryptor.finalize()).decode()
                                              ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 186, in finalize
    data = self._ctx.finalize()
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/cryptography/hazmat/backends/openssl/ciphers.py", line 222, in finalize
    raise ValueError(
ValueError: The length of the provided data is not a multiple of the block length.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 1163, in _read_ready
    self._protocol.datagram_received(data, addr)
  File "/config/custom_components/localtuya/discovery.py", line 67, in datagram_received
    data = data.decode()
           ^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xd2 in position 4: invalid continuation byte
```

Next I'll try wiping LocalTuya from my system and re-installing.

**Comment by amichaelwagner** [comment]
Having the same problem on my end as well

**Comment by BuckinghamIO** [comment]
Same problem here and I have no devices existing currently so this is failing at the discovery phase.

**Comment by mad-tunes** [comment]
Same here, I think.
Swear I search but didnt find this to begin with, logged this in a new issue, then found this one:

The problem
Environment

    Localtuya version: 5.2.1
    Home Assistant Core version: 2024.1.2
    Are you using the Tuya/SmartLife App in parallel ?

Steps to reproduce

    Download LocalTuya through HACS
    Restart HA
    Settings>Devices & Services>Add Integration>Local Tuya
    On Cloud API account configuration screen, change nothing apart from ticking 'Do not configure a Cloud API account', Submit

Feels like I must be missing something basic somewhere?
DP dump
Provide Home Assistant traceback/logs

This error originated from a custom integration.

```
Logger: homeassistant
Source: custom_components/localtuya/discovery.py:67
Integration: LocalTuya integration (documentation, issues)
First occurred: 21:40:48 (3 occurrences)
Last logged: 21:40:51

Error doing job: Exception in callback _SelectorDatagramTransport._read_ready()
Traceback (most recent call last):
  File "/config/custom_components/localtuya/discovery.py", line 65, in datagram_received
    data = decrypt_udp(data)
           ^^^^^^^^^^^^^^^^^
  File "/config/custom_components/localtuya/discovery.py", line 30, in decrypt_udp
    return _unpad(decryptor.update(message) + decryptor.finalize()).decode()
                                              ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 186, in finalize
    data = self._ctx.finalize()
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/cryptography/hazmat/backends/openssl/ciphers.py", line 222, in finalize
    raise ValueError(
ValueError: The length of the provided data is not a multiple of the block length.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 1163, in _read_ready
    self._protocol.datagram_received(data, addr)
  File "/config/custom_components/localtuya/discovery.py", line 67, in datagram_received
    data = data.decode()
           ^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb1 in position 5: invalid start byte
```

Additional information

No devices, no API account set up, similar messages all over HAs logs

**Comment by MrBamse** [comment]
Same here

**Comment by GKG-ctrl** [comment]
Same issue here.. any ideas please

**Comment by EPdeHaas** [comment]
Also encountered this issue, and been looking into it a bit (not the code ... yet), but which devices caused it. And I have managed to isolate the culprits.

Both of these devices are LSC/Battletron devices from the Action (nl). A desk pad and a mouse, They both have firmware that can connect (don't know which route) to the Razer Chroma API. However, this only works when using the branded LSC app, and not with the native tuya app which suggest that they added this functionally on top of the tuya functionality, probably adding or changing the meaning of some of the DPS.

Aside from these 2, which cause these errors, I have 2 more IR blasters that didn't expose there configured buttons to localtuya, so I have a secondary cloud based setup for those anyway. It would be nice though if there was an option to 'blacklist' devices from the auto discovery. This would also mean that the whole discovery chain would not have to run for those devices that you won't add anyway.

And yes. it's a mouse that connects to wifi to control it's rgb.

PS: They are linked to through the native tuya app (I don't use Chroma), just on a different account, and added to HA with the native tuya integration.

PS 2: For now I moved the mouse and desk pad to a guest network so localtuya can't see them.
